chips/litex/src/led_controller.rs

```diff
@@ -0,0 +1,152 @@
+
+use crate::litex_registers::{LiteXSoCRegisterConfiguration, Read, Write};
+
+// TODO: Make the register width adaptable, perhaps by another trait
+// with the integer type as an associated type?
+
+#[repr(C)]
+pub struct LiteXLedRegisters<R: LiteXSoCRegisterConfiguration> {
+    leds_out: R::ReadWrite8,
+}
+
+pub struct LiteXLedController<R: LiteXSoCRegisterConfiguration> {
+    regs: StaticRef<LiteXLedRegisters<R>>,
+    led_count: usize,
+    led_references: Cell<u8>,
+    active_low: bool,
+}
+
+impl<R: LiteXSoCRegisterConfiguration> LiteXLedController<R> {
+    pub fn new(
+        base: StaticRef<LiteXLedRegisters<R>>,
+        led_count: usize,
+        active_low: bool,
+    ) -> LiteXLedController<R> {
+        // The number of leds may not be larger than the bit width of
+        // the supplied register layout
+        //
+        // TODO: Automatically determine based on the type
+        assert!(
+            led_count <= 8,
+            "LiteXLedController register width insufficient to support the requested LED count"
+        );
+
+        LiteXLedController {
+            regs: base,
+            led_count,
+            led_references: Cell::new(0),
+            active_low,
+        }
+    }
+
+    pub fn initialize(&self) {
+        self.regs.leds_out.set(if self.active_low {
+            ((1 << self.led_count() as isize) - 1) as u8
+        } else {
+            0
+        });
+    }
+
+    pub fn led_count(&self) -> usize {
+        self.led_count
+    }
+
+    pub fn get_led<'a>(&'a self, index: usize) -> Option<LiteXLed<'a, R>> {
+        if index < self.led_count() && (self.led_references.get() & (1 << index)) == 0 {
+            self.led_references
+                .set(self.led_references.get() | (1 << index));
+            Some(LiteXLed::new(self, index))
+        } else {
+            None
+        }
+    }
+
+    /// Create a LiteXLed instance referencing a specific LED on the
+    /// controller
+    ///
+    /// This function must only be used in a panic handler, if no
+    /// other code will be running afterwards, in order to guarantee
+    /// consistency between ownership of the LiteXLed instance and
+    /// control over the LED state
+    ///
+    /// This function only checks whether the requested LEDs is within
+    /// the controller's range of available LEDs, but *NOT* whether
+    /// there already is a different reference to the same LED.
+    pub unsafe fn panic_led<'a>(&'a self, index: usize) -> Option<LiteXLed<'a, R>> {
+        if index < self.led_count() {
+            Some(LiteXLed::new(self, index))
+        } else {
+            None
+        }
+    }
+
+    fn destroy_led(&self, index: usize) {
+        self.led_references
+            .set(self.led_references.get() & !(1 << index));
+    }
+
+    fn set_led(&self, index: usize, val: bool) {
+        if val ^ self.active_low {
+            self.regs
+                .leds_out
+                .set(self.regs.leds_out.get() | (1 << index));
```

Comment:
Why not just use `modify()`?

Reply:
`modify()` takes a `FieldValue<T, R>` and this logic should be entirely generic over the register length (as adding more LEDs will cause the register width to simply grow). While I could technically construct a FieldValue with a matching mask, shift and value here, this seems like the more natural and readable option in this case, since I'm effectively only setting single bits in a flat integer.